### PR TITLE
Main window

### PR DESCRIPTION
**Pull request by harrisonyei**
Separate files from QtWidget.
Add StreamMainWindow

### CODE REVIEW

File: src/StreamMainWindow.cpp [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+#include "StreamMainWindow.h"
+#include "Components\Controller.h"
```
Comment (john0312) [inline src/StreamMainWindow.cpp]: Please use "/" for the slash.

File: src/StreamMainWindow.cpp [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+#include <sstream>
+#include <iostream>
+StreamMainWindow::StreamMainWindow(QWidget *parent)
+    : QMainWindow(parent)
+{
+	// setup UIs of the widget.
+    ui.setupUi(this);
```
Comment (john0312) [inline src/StreamMainWindow.cpp]: Please fix incorrect indentation.

File: src/StreamMainWindow.cpp [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+{
+	// handle watch-button clicked event.
+	std::cout << "watch!\n";
+
+	// read input from GUI
+	std::stringstream ss;
+	// remote ip address to watch stream
+	std::string ip;
+	ss << ui.ip_adress_line->text().toStdString();
+	ss >> ip;
```
Comment (john0312) [inline src/StreamMainWindow.cpp]: Please check the user's input.

File: src/StreamMainWindow.cpp [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+	ss.clear();
+	// remote port to listen
+	int listen_port;
+	ss << ui.listen_port_line->text().toStdString();
+	ss >> listen_port;
```
Comment (john0312) [inline src/StreamMainWindow.cpp]: Please check if int parsed successfully.

File: src/StreamMainWindow.cpp [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+
+	// read input from GUI
+	std::stringstream ss;
+	// remote ip address to watch stream
+	std::string ip;
+	ss << ui.ip_adress_line->text().toStdString();
+	ss >> ip;
+	ss.clear();
+	// remote port to listen
+	int listen_port;
```
Comment (john0312) [inline src/StreamMainWindow.cpp]: Please use uint16_t for port.

File: src/StreamMainWindow.h [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+// Handles the interactions between user gui events and controller.
+class StreamMainWindow : public QMainWindow
+{
+    Q_OBJECT
+
+public:
+	// Constructor
+	// @param QWidet* parent
+	// Constructs a qtwidget under the parent qtwidget.
+	// Creates a stream contoller, then initializes it.
```
Comment (john0312) [inline src/StreamMainWindow.h:42]: Please fix indentation.

File: src/StreamMainWindow.h [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+	void onLeaveButtonClicked();
+
+	// Handles Watch-button clicked event.
+	// Starts watching a live stream that is from the given ip and port.
+	void onWatchButtonClicked();
+
+private:
+
+	// UI configurations of this window.
+    Ui::StreamMainWindowClass ui;
```
Comment (john0312) [inline src/StreamMainWindow.h]: Please fix indentation.

File: src/StreamMainWindow.cpp [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+	// read input from GUI
+	std::stringstream ss;
+	// remote ip address to watch stream
+	std::string ip;
+	ss << ui.ip_adress_line->text().toStdString();
+	ss >> ip;
+	ss.clear();
+
+	// Regex expression for validating IPv4
+	std::regex ipv4("(([0-9]|[1-9][0-9]|1[0-9][0-9]|2[0-4][0-9]|25[0-5])\\.){3}([0-9]|[1-9][0-9]|1[0-9][0-9]|2[0-4][0-9]|25[0-5])");
```
Comment (john0312) [inline src/StreamMainWindow.cpp]: Please decide if you want to move this logic out of this class, or add unit test to this class.

**Comment by john0312** [comment]
Please fix the git commit by squashing/rebasing.

**Comment by jeremy5189** [comment]
Please avoid force-pushing, this will cause confusing for the reviewer
Just add another commit if you are trying to fix an issue mentioned here

**Review by jeremy5189 — APPROVED** [review]
It's good for now.
One advice: consider using a linter to uniform coding style, like the position of `{` after `if` statement is different from class

File: src/StreamMainWindow.cpp [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+	// Allocate a stream controller, then initialize it with the widget to be updated.
+	streamController = new Controller();
+	streamController->init(ui.displayGLWidget);
+}
+
+StreamMainWindow::~StreamMainWindow()
+{
+	// stops stream
+	streamController->leave();
+	delete streamController;
```
Comment (john0312) [inline src/StreamMainWindow.cpp]: Consider setting streamController to nullptr after delet.

File: src/StreamMainWindow.cpp (hunk 4 above)
Comment (john0312) [reply src/StreamMainWindow.cpp]: Or add a TODO.

File: src/StreamMainWindow.h [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+	// Handles Watch-button clicked event.
+	// Starts watching a live stream that is from the given ip and port.
+	void onWatchButtonClicked();
+
+private:
+	// UI configurations of this window.
+	Ui::StreamMainWindowClass ui;
+
+	// Controller that controls states of streaming components.
+	Controller* streamController;
```
Comment (john0312) [inline src/StreamMainWindow.h]: Please fix naming.

**Review by john0312 — APPROVED** [review]

**Pull request merged**